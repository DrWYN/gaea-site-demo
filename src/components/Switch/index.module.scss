$width: 50px;
$height: 24px;

.base {
  width: $width;
  height: $height;
  border-radius: $height;
  will-change: background-color;
  transition: background-color .15s linear;
  -webkit-transition: background-color .15s linear;

  > div {
    width: $height;
    height: 100%;
    border-radius: 50%;
    border: 1px solid rgba(0, 0, 0, 0.1);
    position: relative;
    will-change: left;
    transition: left .15s ease-in-out;
    -webkit-transition: left .15s ease-in-out;
  }
}

.on {
  @extend .base;
  border: 1px solid transparent;
  > div {
    left: $width - $height;
  }
}

.off {
  @extend .base;
  border: 1px solid #e7e7e7;
  > div {
    left: 0;
  }
}

.disabled {
  @extend .base;
  border: 1px solid #e7e7e7;
}
