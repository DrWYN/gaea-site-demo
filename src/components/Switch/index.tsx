import * as React from 'react';
import View from '../View/index';
import Touchable from 'Components/Touchable';
import styles from './index.module.scss';

interface ISwitchProps {
  testID?: string;
  className?: string;
  style?: any;
  thumbColor?: string;
  activeThumbColor?: string;
  disabledThumbColor?: string;
  trackColor?: string;
  activeTrackColor?: string;
  disabledTrackColor?: string;
  disabled?: boolean;
  value?: boolean; // 受控，组件的展示受这个值的控制
  onValueChange?: (value: boolean) => void;
}

interface ISwitchState {
  value: boolean
};

interface IStyle {
  width?: number;
  height?: number;
  borderRadius?: number;
  left?: number;
  backgroundColor?: string;
}

const DEFAULT_WIDTH = 50;
const DEFAULT_HEIGHT = 24;

class Switch extends React.Component<ISwitchProps, ISwitchState> {

  constructor(props: ISwitchProps) {
    super(props);
  }

  onValueChange = () => {
    const { value, disabled = false, onValueChange } = this.props;

    if (disabled) {
      return;
    }

    if (typeof onValueChange === 'function') {
      onValueChange(!value);
    }
  }

  render() {
    const {
      className = '',
      thumbColor = '#fff',
      activeThumbColor = '#fff',
      disabledThumbColor = '#fff',
      trackColor = '#fff',
      activeTrackColor = '#5ecb7e',
      disabledTrackColor = '#e7e7e7',
      disabled = false,
      style = {},
      value,
      testID = ''
    } = this.props;
    const { width, height } = style;
    const classNameStr = `${disabled ? styles.disabled : (value ? styles.on : styles.off)} ${className}`;
    const trackStyle: IStyle = {
      backgroundColor: disabled ? disabledTrackColor : (value ? activeTrackColor : trackColor)
    };
    const thumbStyle: IStyle = {
      backgroundColor: disabled ? disabledThumbColor : (value ? activeThumbColor : thumbColor),
      left: disabled ? 0 : (value ? (width || DEFAULT_WIDTH) - (height || DEFAULT_HEIGHT) : 0)
    };

    if (width) {
      trackStyle.width = width;
    }
    if (height) {
      trackStyle.height = height;
      trackStyle.borderRadius = height;
      thumbStyle.width = height;
    }

    return (
      <Touchable onPress = {this.onValueChange}>
      <View
        className = {classNameStr}
        style = {{...style, ...trackStyle}}
        data-testid = {testID}
      >
        <View style={thumbStyle}/>
      </View>
      </Touchable>
    );
  }
}

export default Switch;
