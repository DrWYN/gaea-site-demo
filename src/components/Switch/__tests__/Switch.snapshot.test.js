import React from 'react';
import renderer from 'react-test-renderer';
import Switch from '../index';

describe('<Switch />', () => {

  it('props msg', () => {
    const component = renderer.create(
      <Switch
        onValueChange={(v)=>{console.log(v)}}
        width={100}
        height={100}
        testID={'test'}
      />
    );

    let snapshot = component.toJSON();
    expect(snapshot).toMatchSnapshot();
  });
})
