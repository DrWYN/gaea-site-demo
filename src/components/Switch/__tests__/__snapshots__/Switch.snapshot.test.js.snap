// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> props msg 1`] = `
<div
  className="undefined "
  data-testid=""
  onMouseDown={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "backgroundColor": "#fff",
    }
  }
>
  <div
    data-testid=""
    style={
      Object {
        "backgroundColor": "#fff",
        "left": 0,
      }
    }
  />
</div>
`;
